src/Microsoft.AspNetCore.Mvc.Razor/Internal/DefaultRoslynCompilationService.cs

```diff
@@ -196,7 +198,7 @@ private CSharpCompilation Rewrite(CSharpCompilation compilation)
                 rewrittenTrees.Add(rewrittenTree);
             }
 
-            return compilation.RemoveAllSyntaxTrees().AddSyntaxTrees(rewrittenTrees);
+            compilationContext.Compilation = compilation.RemoveAllSyntaxTrees().AddSyntaxTrees(rewrittenTrees);
```

Comment:
There's a `ReplaceSyntaxTree` (singular) we could do, but ideally that would be part of the loop. The issue there is that we change the semantics of the rewrite ever so slightly (you'll see a rewritten tree if you are looking at other trees as part of the rewritter as opposed to the original tree).

i.e.

```
foreach (var tree in compilation.SyntaxTrees)
{
    var semanticModel = compilation.GetSemanticModel(tree, ignoreAccessibility: true);
    var rewriter = new ExpressionRewriter(semanticModel);

    var rewrittenTree = tree.WithRootAndOptions(rewriter.Visit(tree.GetRoot()), tree.Options);
    compilation = compilation.ReplaceSyntaxTree(tree, rewrittenTree);
}
```